doc/md/cloud: setting up ci with atlas cloud guide (#1970)
ariga/atlas: doc/md/cloud/setup-ci.mdx

---
id: setup-ci
title: Setting up CI for your Database with Atlas Cloud
---

As your application evolves, so does your database schema. If you're not careful

> with schema changes (migrations) you can end up introducing all sorts of issues

Reviewer: ```suggestion
with schema changes (migrations), you can end up introducing all sorts of issues
```

that are painful and expensive to fix.

To mitigate the risk of deploying dangerous changes to database schemas, many teams
apply CI/CD practices to their database. This means that every change to the database
schema is automatically reviewed and tested before it is deployed to production.

In this guide, we'll show you how to set up CI for your database using Atlas Cloud.

## Push or Pull Workflows

There are two ways to set up CI for your database with Atlas Cloud:

* **Pull (GitHub App)** - You install the Atlas Cloud GitHub App on your GitHub

>   repository and Atlas Cloud will automatically pull changes from your GitHub

Reviewer: ```suggestion
  repository. Atlas Cloud will automatically pull changes from your GitHub
```

Reviewer: Breaking up a long sentence makes it easier to read. Otherwise, the sentence is fine.

>   repository and run tests on them.

Reviewer: ```suggestion
  repository, then run tests on them.
```

Reviewer: Again, easier to read. Gets rid of too many "and"s

  This is the fastest way to set Atlas Cloud up, but it does require that you
  have the necessary permissions to install the GitHub App on your repository.

* **Push (GitHub Action)** - You install the [`ariga/atlas-action`](https://github.com/ariga/atlas-action)
  GitHub Action on your GitHub repository and configure it to run on any pull request
  that modifies your database schema.

>   This is a bit more work to set up, but it doesn't require you to grant Atlas

Reviewer: ```suggestion
  This is a bit more work to set up, but it does _not_ require you to grant Atlas
```

Reviewer: Emphasizes difference with first option

  Cloud any permissions on your repository.

## Prerequisites

1. A GitHub repository containing an Atlas migration directory. If you don't have one handy

>    and would like to experiment with this workflow you can use the

Reviewer: ```suggestion
   and would like to experiment with this workflow, you can use the
```

   [`ariga/atlas-template`](https://github.com/ariga/atlas-template) template repo
   to create one for you.
2. An [Atlas Cloud](https://atlasgo.cloud) account. If you don't have one, you can
   [sign up for free](https://auth.atlasgo.cloud/signup).

## Pull (GitHub App) Workflow

### Step 1: Install the Atlas Cloud GitHub App

After logging in to Atlas Cloud, create a new migration directory by clicking on
"Migrations Directories" in the sidebar and then clicking on the "Add Directory"
button.

![](https://atlasgo.io/uploads/cloud/ci-guide/create-dir-1.png)

Click "Connect with GitHub" to initiate the GitHub App installation flow.

![](https://atlasgo.io/uploads/cloud/ci-guide/create-dir-2.png)

### Step 2: Connect Your Repository

Once you've installed the GitHub App, you need to grant it access to
the relevant repositories. Click on the "Organization" field and choose
"Add Organization" from the dropdown:

![](https://atlasgo.io/uploads/cloud/ci-guide/create-dir-3.png)

A popup requesting you to grant permission to specific repositories will appear.

> Select the repositories you want to grant access to and click "Install and Authorize".

Reviewer: ```suggestion
Select the repositories you want to grant access to, then click "Install and Authorize".
```

![](https://atlasgo.io/uploads/cloud/ci-guide/create-dir-4.png)

### Step 3: Configure the Migration Directory

Fill the rest of the form with the relevant configuration details:
* **Directory Path** - The path to the directory containing your migrations relative to your repository root.
* **Directory Name** - An identifier for the directory. This will be used to identify the directory in the Atlas Cloud and APIs.
* **Migration Engine** - The format of the migrations in the directory. Atlas works best with migration directories in its

>   format, but also supports `golang-migrate`, Flyway and Liquibase.

Reviewer: ```suggestion
  format, but also supports `golang-migrate`, Flyway, and Liquibase.
```

results of the CI run:

![](https://atlasgo.io/uploads/cloud/ci-guide/app-ci.png)

Amazing! Atlas Cloud caught the destructive change and prevented it from being
accidentally merged to your main branch.

To see a more detailed report of the CI run, click on the "Full Report on Atlas Cloud"
link in the comment:

![](https://atlasgo.io/uploads/cloud/ci-guide/app-cloud-report.png)

## Push (GitHub Action) Workflow

### Step 1: Create a Bot Token for Atlas Cloud

In order to report the results of your CI runs to Atlas Cloud, you will need to
create a bot token for Atlas Cloud to use. Follow [these instructions](/bot.mdx)
to create a token and save it somewhere safe.

### Step 2: Install the Atlas GitHub CLI Extension

To streamline the process of configuring the GitHub Action, we've created a
GitHub CLI extension that will do most of the work for you:

1. Make sure you have the GitHub CLI installed.

   ```shell
   brew install gh
   ```
   See [here](https://cli.github.com/manual/installation) for more installation options.

2. Install the Atlas GitHub CLI extension:

   ```shell
   gh extension install ariga/atlas
   ```

### Step 3: Configure the GitHub Action

> 1. Make sure you have the necessary permissions to configure your action:

Reviewer: ```suggestion
1. **Permissions** - Make sure you have the necessary permissions to configure your action:
```

Reviewer: I find that making the list of steps easy to remember makes absorbing the details of each list item much easier later. Hence, I made the list items 1. Permissions, 2. Root, 3. Configure so that anyone reading the tutorial can keep track of the scope of what they're doing.

  ```shell
  gh auth refresh -s write:packages,workflow
  ```

> 2. Make sure your current working directory is the root of your repository.

Reviewer: ```suggestion
2. **Root** - Make sure your current working directory is the root of your repository.
```

  ```shell
  cd path/to/my/repo
  ```

> 3. Run the `init-ci` command to configure the GitHub Action:

Reviewer: ```suggestion
3. **Configure** - Run the `init-ci` command to configure the GitHub Action:
```

  ```shell
  gh atlas init-action --token <your-bot-token>
  ```
  Atlas will scan your repository (locally) for directories containing Atlas migrations
  and ask you which one you would like to use for CI. Select the desired directory and press "Enter":
  ```text
  Use the arrow keys to navigate: ↓ ↑ → ←
  ? choose migration directory:
    ▸ migrations
  ```

>   Atlas will then which database driver this directory contains migrations for. Select the

Reviewer: ```suggestion
  Atlas will then ? which database driver this directory contains migrations for. Select the
```
Missing word?
